### PR TITLE
plugin new now takes a PluginInit<T> struct (generic over config)

### PR DESCRIPTION
**Pull request by garypen**
After discussion with router team, the preferred direction for exposing a config was agreed to be via the existing `new` method. This method only expected a plugin specific config struct, so this change is mainly about modifying that so that all plugins now expect a `PluginInit` struct which is generic over plugin config.

The first use of this new capability is to pass in the router supergraph schema as part of `PluginInit`.

fixes: #1415

### CODE REVIEW

File: apollo-router/src/plugin/mod.rs [hunk 1]
```diff
@@ -48,10 +48,33 @@ use crate::RouterResponse;
...
+    fn(&serde_json::Value, String) -> BoxFuture<Result<Box<dyn DynPlugin>, BoxError>>;
 
 type SchemaFactory = fn(&mut SchemaGenerator) -> schemars::schema::Schema;
 
+/// Initialise details for a plugin
+pub struct PluginInit<T> {
+    /// Configuration
+    pub config: T,
+    /// Router Supergraph Schema
+    pub schema: String,
```
Comment (BrynCooke) [inline apollo-router/src/plugin/mod.rs:61]: Maybe make this non-pub and add an accessor?
That way when apollo-rs is introduced maybe the string schema can just be derived.

**Review by BrynCooke — APPROVED** [review]
LGTM

File: apollo-router/src/plugin/mod.rs (hunk 1 above)
Comment (SimonSapin) [reply apollo-router/src/plugin/mod.rs]: +1

Also having at least one private field allows us to add more fields later without that being a breaking change

File: apollo-router/src/plugin/mod.rs [hunk 2]
```diff
@@ -48,10 +48,33 @@ use crate::RouterResponse;
...
+    T: for<'de> Deserialize<'de>,
+{
+    pub fn new(config: T, schema: String) -> Self {
+        PluginInit { config, schema }
+    }
+
+    pub fn try_new(config: serde_json::Value, schema: String) -> Result<Self, BoxError> {
+        let config: T = serde_json::from_value(config)?;
+        Ok(PluginInit { config, schema })
+    }
```
Comment (SimonSapin) [inline apollo-router/src/plugin/mod.rs:81]: Could these be `pub(crate)` instead?

Comment (garypen) [reply apollo-router/src/plugin/mod.rs:81]: `new` is used in the integration tests, so I can't make that `pub(crate)`. `try_new` is used in the new `supergraph_sdl` example. Arguably I could use `new` in the example, but it's more convenient to use `try_new`, so I think they both need to be pub.

File: apollo-router/src/plugin/mod.rs (hunk 1 above)
Comment (garypen) [reply apollo-router/src/plugin/mod.rs]: I'm going to leave this as is for now and we can consider if we want to change things as part of the 1.0 decision making.

**Pull request merged**